### PR TITLE
fix: unused index on jsonb/jsonb arrow filter

### PR DESCRIPTION
**Pull request by steve-chavez**
Closes https://github.com/PostgREST/postgrest/issues/2594

Before:
```
curl "localhost:3000/bets?data_jsonb->>contractId=eq.1" -H "Accept: application/vnd.pgrst.plan"

Aggregate  (cost=32.26..32.28 rows=1 width=112)
  ->  Seq Scan on bets  (cost=0.00..32.23 rows=6 width=36)
        Filter: ((to_jsonb(data_jsonb) ->> 'contractId'::text) = '1'::text)
```

Now:

```
curl "localhost:3000/bets?data_jsonb->>contractId=eq.1" -H "Accept: application/vnd.pgrst.plan"

Aggregate  (cost=12.67..12.69 rows=1 width=112)
  ->  Bitmap Heap Scan on bets  (cost=4.18..12.65 rows=4 width=68)
        Recheck Cond: ((data_jsonb ->> 'contractId'::text) = '1'::text)
        ->  Bitmap Index Scan on bets_data_jsonb  (cost=0.00..4.18 rows=4 width=0)
              Index Cond: ((data_jsonb ->> 'contractId'::text) = '1'::text)
```

### CODE REVIEW

File: src/PostgREST/Query/SqlFragment.hs [hunk 1]
```diff
@@ -241,10 +241,13 @@ pgFmtCallUnary :: Text -> SQL.Snippet -> SQL.Snippet
...
--- Using to_jsonb instead of to_json to avoid missing operator errors when filtering:
--- "operator does not exist: json = unknown"
-pgFmtField table CoercibleField{cfName=fn, cfJsonPath=jp} = "to_jsonb(" <> pgFmtColumn table fn <> ")" <> pgFmtJsonPath jp
+pgFmtField table CoercibleField{cfName=fn, cfJsonPath=[]}                = pgFmtColumn table fn
+pgFmtField table CoercibleField{cfName=fn, cfIRType=typ, cfJsonPath=jp}  =
+  case typ of
+    "json"  -> pgFmtColumn table fn <> pgFmtJsonPath jp
+    "jsonb" -> pgFmtColumn table fn <> pgFmtJsonPath jp
+    -- Using to_jsonb instead of to_json to avoid missing operator errors when filtering: "operator does not exist: json = unknown"
+    _       -> "to_jsonb(" <> pgFmtColumn table fn <> ")" <> pgFmtJsonPath jp
```
Comment (steve-chavez) [inline src/PostgREST/Query/SqlFragment.hs:250]: This wasn't right. There's another logic that converts the `cfIRType` to `json` here: 

https://github.com/PostgREST/postgrest/blob/e332f038efb46f55d0f463d5ada2f6a812b6410d/src/PostgREST/Plan.hs#L229-L235

It's a bit confusing to have the logic split in two places. Will try to refactor this.

Comment (steve-chavez) [reply src/PostgREST/Query/SqlFragment.hs:250]: Fixing applying the index for order is also another process. Currently not working.

File: test/spec/Feature/Query/PlanSpec.hs [hunk 2]
```diff
@@ -348,6 +348,43 @@ spec actualPgVersion = do
...
+          resBody `shouldSatisfy` (\t -> T.isInfixOf "Index" (decodeUtf8 $ LBS.toStrict t))
+
+      it "should use an index for ordering on a jsonb arrow operator" $ do
+        r <- request methodGet "/bets?order=data_jsonb->>contractId"
+               [(hAccept, "application/vnd.pgrst.plan")] ""
+
+        let resBody = simpleBody r
+
+        liftIO $ do
+          resBody `shouldSatisfy` (\t -> T.isInfixOf "Index" (decodeUtf8 $ LBS.toStrict t))
```
Comment (steve-chavez) [inline test/spec/Feature/Query/PlanSpec.hs:386]: To fix these I had to add a new CoercibleOrderTerm type. This is so we know the type of the order field and not apply `to_jsonb` unnecessarily.

File: src/PostgREST/Plan.hs [hunk 3]
```diff
@@ -228,11 +228,14 @@ resolveTableFieldName table fieldName =
...
 resolveTableField table (fieldName, jp) =
   case resolveTableFieldName table fieldName of
-    cf@CoercibleField{cfIRType=""} -> cf{cfJsonPath=jp}
-    cf                             -> cf{cfJsonPath=jp, cfIRType="json"}
+    -- types that are already json/jsonb don't need to be converted with `to_jsonb` for using arrow operators `data->attr`
+    -- this prevents indexes not applying https://github.com/PostgREST/postgrest/issues/2594
+    cf@CoercibleField{cfIRType="json"}  -> cf{cfJsonPath=jp}
+    cf@CoercibleField{cfIRType="jsonb"} -> cf{cfJsonPath=jp}
+    -- other types will get converted `to_jsonb(col)->attr`
+    cf                                  -> cf{cfJsonPath=jp, cfToJson=True}
```
Comment (steve-chavez) [inline src/PostgREST/Plan.hs:238]: This is the main logic for the fix. Now is in only one place.

**Review by laurenceisla — APPROVED** [review]

**Pull request merged**